core/cloudimagemetadata/types.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package cloudimagemetadata
+
+import "time"
+
+// MetadataAttributes contains cloud image metadata attributes.
+type MetadataAttributes struct {
+	// Stream contains reference to a particular stream,
+	// for e.g. "daily" or "released"
+	Stream string
+
+	// Region is the name of cloud region associated with the image.
+	Region string
+
+	// Version is OS version, for e.g. "22.04".
+	Version string
+
+	// Arch represents the architecture of the cloud image, for example, "amd64" or "arm64".
+	Arch string
+
+	// VirtType contains virtualisation type of the cloud image, for e.g. "pv", "hvm". "kvm".
+	VirtType string
+
+	// RootStorageType contains type of root storage, for e.g. "ebs", "instance".
+	RootStorageType string
+
+	// RootStorageSize contains size of root storage in gigabytes (GB).
+	RootStorageSize *uint64
+
+	// Source describes where this image is coming from: is it public? custom?
+	Source string
+}
+
+// Metadata describes a cloud image metadata.
+type Metadata struct {
+	MetadataAttributes
+
+	// Priority is an importance factor for image metadata.
+	// Higher number means higher priority.
+	// This will allow to sort metadata by importance.
+	Priority int
+
+	// ImageId contains image identifier.
+	ImageId string
+
+	// CreationTime contains the time and date the image was created. This
+	// is populated when the Metadata is saved.
+	CreationTime time.Time
+}
+
+// MetadataFilter contains all metadata attributes that allows to find a particular
+// cloud image metadata. Since size and source are not discriminating attributes
+// for cloud image metadata, they are not included in search criteria.
+type MetadataFilter struct {
+	// Region stores metadata region.
+	Region string `json:"region,omitempty"`
```

Comment:
Remove the `json` tag, these shouldn't be in core. The core package _should_ be clean from wire transfer protocols. We should serialize in the API server to the core filter type. We can then use that everywhere.